PlayFabSDK/source/Qos/RegionPinger.cs

```diff
@@ -3,76 +3,136 @@ namespace PlayFab.QoS
             {
-                if (!measure.IsCompleted || measure.Result < 0)
+                ct.Cancel();
+                int pingResultMs = await pingResultTask;
+
+                lock (latencyMeasures)
                 {
-                    errorCode = (int) QosErrorCode.Timeout;
-                    continue;
+                    latencyMeasures.Add(pingResultMs);
                 }
+            }
+            else
+            {
+                Interlocked.Increment(ref _numTimeouts);
+            }
+        }
 
-                latencies.Add(measure.Result);
+        public QosRegionResult GetResult()
+        {
+            if (IsAtTimeoutThreshold())
+            {
+                return new QosRegionResult
+                {
+                    Region = _region,
+                    LatencyMs = UnknownLatencyValue,
+                    RawMeasurements = latencyMeasures,
+                    NumTimeouts = _numTimeouts,
+                    ErrorCode = (int) QosErrorCode.Timeout
+                };
             }
 
-            int averageLatency = UnknownLatencyValue;
-            if (latencies.Count > 0)
+            int averageLatency;
+            QosErrorCode errorCode;
+
+            lock (latencyMeasures)
             {
-                long sum = 0;
-                foreach (int latency in latencies)
+                if (latencyMeasures.Count > 0)
                 {
-                    sum += latency;
-                }
+                    long sum = 0;
+                    int count = 0;
+                    if (latencyMeasures.Count >= 3)
+                    {
+                        // If there are at least 4, throw out the top and bottom measurements 
+                        latencyMeasures.Sort();
+                        for (int i = 1; i < latencyMeasures.Count - 1; i++)
+                        {
+                            count++;
+                            sum += latencyMeasures[i];
+                        }
+                    }
+                    else
+                    {
+                        for (int i = 0; i < latencyMeasures.Count; i++)
+                        {
+                            count++;
+                            sum += latencyMeasures[i];
+                        }
+                    }
 
-                averageLatency = (int)(sum / latencies.Count);
+                    errorCode = QosErrorCode.Success;
+                    averageLatency = (int) (sum / count);
+                }
+                else
+                {
+                    errorCode = QosErrorCode.NoResult;
+                    averageLatency = UnknownLatencyValue;
+                }
             }
 
-            // Return the average of the remaining numbers
             return new QosRegionResult
             {
                 Region = _region,
                 LatencyMs = averageLatency,
-                ErrorCode = errorCode
+                RawMeasurements = latencyMeasures,
```

Comment:
It'll also make it clear when you're chopping values, and would have made it more visible if you tested 3, and chopped to 1

Reply:
It is intentionally not the chopped value. The idea being that it reports all values in case developers would like to do their own aggregation function.

---
In reply to: [376149778](https://github.com/PlayFab/CSharpSDK/pull/111#discussion_r376149778) [](ancestors = 376149778)

Reply:
I slightly mis-read this the first time.
On re-read, and your response, yeah, looks good.